### PR TITLE
Support project definition V2 in streamlit deploy command

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Added support for project definition schema version 2 in `snow streamlit deploy`.  This change requires V2 flag to be enabled.

### CODE REVIEW

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 1]
```diff
@@ -189,6 +189,9 @@ def _convert_env(
             return env
         return ProjectEnvironment(default_env=(env or {}), override_env={})
 
+    def get_entities_by_type(self, entity_type: str):
+        return {i: e for i, e in self.entities.items() if e.get_type() == entity_type}
+
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/project_definition.py:194]: @sfc-gh-jsikorski you may be interested in this helper.

File: src/snowflake/cli/plugins/streamlit/commands.py [hunk 2]
```diff
@@ -134,37 +139,51 @@ def streamlit_deploy(
...
+        if not artefact.exists():
+            raise ClickException(
+                f"Specified artefact {artefact} does not exist locally."
+            )
+
+    # Validate that main file is in artefacts
+    if Path(streamlit.main_file) not in streamlit.artifacts:
+        raise ClickException(
+            f"Specified main file {streamlit.main_file} is not included in artifacts."
+        )
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/plugins/streamlit/commands.py:175]: Maybe extract validation into separate function. WDYT?

File: src/snowflake/cli/plugins/streamlit/commands.py [hunk 3]
```diff
@@ -173,6 +192,59 @@ def streamlit_deploy(
...
+                "query_warehouse": pd.streamlit.query_warehouse,
+                "main_file": str(pd.streamlit.main_file),
+                "pages_dir": str(pd.streamlit.pages_dir),
+                "stage": pd.streamlit.stage,
+                "artifacts": artefacts,
+            }
+        },
+    }
+    pd = ProjectDefinitionV2(**data)
+    return pd
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/plugins/streamlit/commands.py:245]:
```suggestion
    return ProjectDefinitionV2(**data)
```

File: src/snowflake/cli/plugins/streamlit/commands.py (hunk 2 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/plugins/streamlit/commands.py:175]: Good point, but I would prefer to do this in next PR where I will add support for multiple instances

**Review by sfc-gh-astus — APPROVED** [review]

File: src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    main_file: Optional[str] = Field(
+        title="Entrypoint file of the Streamlit app", default="streamlit_app.py"
+    )
+    pages_dir: Optional[str] = Field(title="Streamlit pages", default=None)
+    stage: Optional[str] = Field(
+        title="Stage in which the app’s artifacts will be stored", default="streamlit"
+    )
+    # Possibly can be PathMapping
+    artifacts: Optional[List[Path]] = Field(
+        title="List of additional files which should be included into deployment artifacts",
```
Comment (sfc-gh-pczajka) [inline src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py:41]: Should we change the description if we require "main_file" to be a part of it?

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py:41]: Good point, will add it

File: src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        title="Entrypoint file of the Streamlit app", default="streamlit_app.py"
+    )
+    pages_dir: Optional[str] = Field(title="Streamlit pages", default=None)
+    stage: Optional[str] = Field(
+        title="Stage in which the app’s artifacts will be stored", default="streamlit"
+    )
+    # Possibly can be PathMapping
+    artifacts: Optional[List[Path]] = Field(
+        title="List of files which should be deployed. Each file needs exist locally. "
+        "Main file needs be included in artifacts.",
```
Comment (sfc-gh-pczajka) [inline src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py:42]:
```suggestion
        title="List of files which should be deployed. Each file needs to exist locally. "
        "Main file needs to be included in the artifacts.",
```

File: src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    )
+    main_file: Optional[str] = Field(
+        title="Entrypoint file of the Streamlit app", default="streamlit_app.py"
+    )
+    pages_dir: Optional[str] = Field(title="Streamlit pages", default=None)
+    stage: Optional[str] = Field(
+        title="Stage in which the app’s artifacts will be stored", default="streamlit"
+    )
+    # Possibly can be PathMapping
+    artifacts: Optional[List[Path]] = Field(
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py:40]: We need to pull up the definition of `stage` and `artifact` so the fields can be shared between Streamlit and Application Package (and likely function / spro).

File: src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                f"Specified main file {self.main_file} is not included in artifacts."
+            )
+        return self
+
+    @model_validator(mode="after")
+    def artifacts_must_exists(self):
+        if not self.artifacts:
+            return self
+
+        for artefact in self.artifacts:
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py:62]: `artifact`

File: src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            if not artefact.exists():
+                raise ValueError(
+                    f"Specified artefact {artefact} does not exist locally."
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/entities/streamlit_entity.py:65]: `artifact`

**Pull request merged**